File: output/rust/tests/32_doc_lamp_hsm.rs

// Documentation Example: HSM Lamp with color behavior factored out


use std::collections::HashMap;

struct LampHSMFrameEvent {
    message: String,
    parameters: std::collections::HashMap<String, Box<dyn std::any::Any>>,
}

impl Clone for LampHSMFrameEvent {
    fn clone(&self) -> Self {
        Self {
            message: self.message.clone(),
            parameters: std::collections::HashMap::new(),
        }
    }
}

impl LampHSMFrameEvent {
    fn new(message: &str) -> Self {
        Self {
            message: message.to_string(),
            parameters: std::collections::HashMap::new(),
        }
    }
}

struct LampHSMFrameContext {
    event: LampHSMFrameEvent,
    _return: Option<Box<dyn std::any::Any>>,
    _data: std::collections::HashMap<String, Box<dyn std::any::Any>>,
}

impl LampHSMFrameContext {
    fn new(event: LampHSMFrameEvent, default_return: Option<Box<dyn std::any::Any>>) -> Self {
        Self {
            event,
            _return: default_return,
            _data: std::collections::HashMap::new(),
        }
    }
}

#[derive(Clone)]
struct LampHSMCompartment {
    state: String,
    state_vars: std::collections::HashMap<String, i32>,
    forward_event: Option<LampHSMFrameEvent>,
}

impl LampHSMCompartment {
    fn new(state: &str) -> Self {
        Self {
            state: state.to_string(),
            state_vars: std::collections::HashMap::new(),
            forward_event: None,
        }
    }
}

#[derive(Clone)]
enum LampHSMStateContext {
    Off,
    On,
    ColorBehavior,
    Empty,
}

impl Default for LampHSMStateContext {
    fn default() -> Self {
        LampHSMStateContext::Off
    }
}

pub struct LampHSM {
    _state_stack: Vec<(String, LampHSMStateContext)>,
    __compartment: LampHSMCompartment,
    __next_compartment: Option<LampHSMCompartment>,
    _context_stack: Vec<LampHSMFrameContext>,
    color: String,
    lamp_on: bool,
}

impl LampHSM {
    pub fn new() -> Self {
        let mut this = Self {
            _state_stack: vec![],
            _context_stack: vec![],
            color: String::from("white"),
            lamp_on: false,
            __compartment: LampHSMCompartment::new("Off"),
            __next_compartment: None,
        };
let __frame_event = LampHSMFrameEvent::new("$>");
this.__kernel(__frame_event);
        this
    }

    fn __kernel(&mut self, __e: LampHSMFrameEvent) {
// Route event to current state
self.__router(&__e);
// Process any pending transition
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    // Exit current state
    let exit_event = LampHSMFrameEvent::new("<$");
    self.__router(&exit_event);
    // Switch to new compartment
    self.__compartment = next_compartment;
    // Enter new state (or forward event)
    if self.__compartment.forward_event.is_none() {
        let enter_event = LampHSMFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        // Forward event to new state
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            // Forwarding enter event - just send it
            self.__router(&forward_event);
        } else {
            // Forwarding other event - send $> first, then forward
            let enter_event = LampHSMFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
    }

    fn __router(&mut self, __e: &LampHSMFrameEvent) {
match self.__compartment.state.as_str() {
    "Off" => self._state_Off(__e),
    "On" => self._state_On(__e),
    "ColorBehavior" => self._state_ColorBehavior(__e),
    _ => {}
}
    }

    fn __transition(&mut self, next_compartment: LampHSMCompartment) {
self.__next_compartment = Some(next_compartment);
    }

    fn _state_stack_push(&mut self) {
let state_context = match self.__compartment.state.as_str() {
    "Off" => LampHSMStateContext::Off,
    "On" => LampHSMStateContext::On,
    "ColorBehavior" => LampHSMStateContext::ColorBehavior,
    _ => LampHSMStateContext::Empty,
};
self._state_stack.push((self.__compartment.state.clone(), state_context));
    }

    fn _state_stack_pop(&mut self) {
let (saved_state, state_context) = self._state_stack.pop().unwrap();
let exit_event = LampHSMFrameEvent::new("$<");
self.__router(&exit_event);
self.__compartment.state = saved_state;
match state_context {
    LampHSMStateContext::Off => {}
    LampHSMStateContext::On => {}
    LampHSMStateContext::ColorBehavior => {}
    LampHSMStateContext::Empty => {}
}
    }

    pub fn turn_on(&mut self) {
let mut __e = LampHSMFrameEvent::new("turn_on");
let __ctx = LampHSMFrameContext::new(__e.clone(), None);
self._context_stack.push(__ctx);
match self.__compartment.state.as_str() {
            "Off" => { self._s_Off_turn_on(&__e); }
            _ => {}
        }
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    let exit_event = LampHSMFrameEvent::new("<$");
    self.__router(&exit_event);
    self.__compartment = next_compartment;
    if self.__compartment.forward_event.is_none() {
        let enter_event = LampHSMFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            self.__router(&forward_event);
        } else {
            let enter_event = LampHSMFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
self._context_stack.pop();
    }

    pub fn turn_off(&mut self) {
let mut __e = LampHSMFrameEvent::new("turn_off");
let __ctx = LampHSMFrameContext::new(__e.clone(), None);
self._context_stack.push(__ctx);
match self.__compartment.state.as_str() {
            "On" => { self._s_On_turn_off(&__e); }
            _ => {}
        }
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    let exit_event = LampHSMFrameEvent::new("<$");
    self.__router(&exit_event);
    self.__compartment = next_compartment;
    if self.__compartment.forward_event.is_none() {
        let enter_event = LampHSMFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            self.__router(&forward_event);
        } else {
            let enter_event = LampHSMFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
self._context_stack.pop();
    }

    pub fn get_color(&mut self) -> String {
let mut __e = LampHSMFrameEvent::new("get_color");
let __ctx = LampHSMFrameContext::new(__e.clone(), None);
self._context_stack.push(__ctx);
match self.__compartment.state.as_str() {
            "Off" => { self._s_ColorBehavior_get_color(&__e); }
            "On" => { self._s_ColorBehavior_get_color(&__e); }
            "ColorBehavior" => { self._s_ColorBehavior_get_color(&__e); }
            _ => {}
        }
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    let exit_event = LampHSMFrameEvent::new("<$");
    self.__router(&exit_event);
    self.__compartment = next_compartment;
    if self.__compartment.forward_event.is_none() {
        let enter_event = LampHSMFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            self.__router(&forward_event);
        } else {
            let enter_event = LampHSMFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
let __ctx = self._context_stack.pop().unwrap();
if let Some(ret) = __ctx._return {
    *ret.downcast::<String>().unwrap()
} else {
    Default::default()
}
    }

    pub fn set_color(&mut self, color: String) {
let mut __e = LampHSMFrameEvent::new("set_color");
__e.parameters.insert("color".to_string(), Box::new(color) as Box<dyn std::any::Any>);
let __ctx = LampHSMFrameContext::new(__e.clone(), None);
self._context_stack.push(__ctx);
match self.__compartment.state.as_str() {
            "Off" => { self._s_ColorBehavior_set_color(&__e, color); }
            "On" => { self._s_ColorBehavior_set_color(&__e, color); }
            "ColorBehavior" => { self._s_ColorBehavior_set_color(&__e, color); }
            _ => {}
        }
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    let exit_event = LampHSMFrameEvent::new("<$");
    self.__router(&exit_event);
    self.__compartment = next_compartment;
    if self.__compartment.forward_event.is_none() {
        let enter_event = LampHSMFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            self.__router(&forward_event);
        } else {
            let enter_event = LampHSMFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
self._context_stack.pop();
    }

    pub fn is_lamp_on(&mut self) -> bool {
let mut __e = LampHSMFrameEvent::new("is_lamp_on");
let __ctx = LampHSMFrameContext::new(__e.clone(), None);
self._context_stack.push(__ctx);
match self.__compartment.state.as_str() {
            "Off" => { self._s_Off_is_lamp_on(&__e); }
            "On" => { self._s_On_is_lamp_on(&__e); }
            _ => {}
        }
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    let exit_event = LampHSMFrameEvent::new("<$");
    self.__router(&exit_event);
    self.__compartment = next_compartment;
    if self.__compartment.forward_event.is_none() {
        let enter_event = LampHSMFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            self.__router(&forward_event);
        } else {
            let enter_event = LampHSMFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
let __ctx = self._context_stack.pop().unwrap();
if let Some(ret) = __ctx._return {
    *ret.downcast::<bool>().unwrap()
} else {
    Default::default()
}
    }

    fn _state_ColorBehavior(&mut self, __e: &LampHSMFrameEvent) {
match __e.message.as_str() {
    "get_color" => { self._s_ColorBehavior_get_color(__e); }
    _ => {}
}
    }

    fn _state_Off(&mut self, __e: &LampHSMFrameEvent) {
match __e.message.as_str() {
    "is_lamp_on" => { self._s_Off_is_lamp_on(__e); }
    "turn_on" => { self._s_Off_turn_on(__e); }
    _ => self._state_ColorBehavior(__e),
}
    }

    fn _state_On(&mut self, __e: &LampHSMFrameEvent) {
match __e.message.as_str() {
    "<$" => { self._s_On_exit(__e); }
    "$>" => { self._s_On_enter(__e); }
    "is_lamp_on" => { self._s_On_is_lamp_on(__e); }
    "turn_off" => { self._s_On_turn_off(__e); }
    _ => self._state_ColorBehavior(__e),
}
    }

    fn _s_ColorBehavior_set_color(&mut self, __e: &LampHSMFrameEvent, color: String) {
self.color = color;
    }

    fn _s_ColorBehavior_get_color(&mut self, __e: &LampHSMFrameEvent) {
if let Some(ctx) = self._context_stack.last_mut() { ctx._return = Some(Box::new(self.color.clone())); }
return;
    }

    fn _s_Off_is_lamp_on(&mut self, __e: &LampHSMFrameEvent) {
if let Some(ctx) = self._context_stack.last_mut() { ctx._return = Some(Box::new(self.lamp_on)); }
return;
    }

    fn _s_Off_turn_on(&mut self, __e: &LampHSMFrameEvent) {
self.__transition(LampHSMCompartment::new("On"));
    }

    fn _s_On_is_lamp_on(&mut self, __e: &LampHSMFrameEvent) {
if let Some(ctx) = self._context_stack.last_mut() { ctx._return = Some(Box::new(self.lamp_on)); }
return;
    }

    fn _s_On_turn_off(&mut self, __e: &LampHSMFrameEvent) {
self.__transition(LampHSMCompartment::new("Off"));
    }

    fn _s_On_enter(&mut self, __e: &LampHSMFrameEvent) {
self.turn_on_lamp();
    }

    fn _s_On_exit(&mut self, __e: &LampHSMFrameEvent) {
self.turn_off_lamp();
    }

    fn turn_on_lamp(&mut self) {
            self.lamp_on = true;
    }

    fn turn_off_lamp(&mut self) {
            self.lamp_on = false;
    }
}


fn main() {
    println!("=== Test 32: Doc Lamp HSM ===");
    let mut lamp = LampHSM::new();

    // Color behavior available in both states
    assert_eq!(lamp.get_color(), "white", "Expected 'white'");
    lamp.set_color(String::from("red"));
    assert_eq!(lamp.get_color(), "red", "Expected 'red'");

    // Turn on
    lamp.turn_on();
    assert!(lamp.is_lamp_on(), "Lamp should be on");

    // Color still works when on
    lamp.set_color(String::from("green"));
    assert_eq!(lamp.get_color(), "green", "Expected 'green'");

    // Turn off
    lamp.turn_off();
    assert!(!lamp.is_lamp_on(), "Lamp should be off");

    // Color still works when off
    assert_eq!(lamp.get_color(), "green", "Expected 'green'");

    println!("PASS: HSM lamp works correctly");
}

File: output/cpp/tests/32_doc_lamp_hsm.cpp
#include <string>
#include <unordered_map>
#include <vector>
#include <any>
#include <memory>
#include <functional>


#include <iostream>
#include <string>
#include <cassert>

// Documentation Example: HSM Lamp with color behavior factored out

class LampHSMFrameEvent {
public:
    std::string _message;
    std::unordered_map<std::string, std::any> _parameters;

    LampHSMFrameEvent(const std::string& message, std::unordered_map<std::string, std::any> params = {})
        : _message(message), _parameters(std::move(params)) {}
};

class LampHSMFrameContext {
public:
    LampHSMFrameEvent _event;
    std::any _return;
    std::unordered_map<std::string, std::any> _data;

    LampHSMFrameContext(LampHSMFrameEvent event, std::any default_return = {})
        : _event(std::move(event)), _return(std::move(default_return)) {}
};

class LampHSMCompartment {
public:
    std::string state;
    std::unordered_map<std::string, std::any> state_args;
    std::unordered_map<std::string, std::any> state_vars;
    std::unordered_map<std::string, std::any> enter_args;
    std::unordered_map<std::string, std::any> exit_args;
    std::unique_ptr<LampHSMFrameEvent> forward_event;
    std::unique_ptr<LampHSMCompartment> parent_compartment;

    explicit LampHSMCompartment(const std::string& state) : state(state) {}

    std::unique_ptr<LampHSMCompartment> clone() const {
        auto c = std::make_unique<LampHSMCompartment>(state);
        c->state_args = state_args;
        c->state_vars = state_vars;
        c->enter_args = enter_args;
        c->exit_args = exit_args;
        return c;
    }
};

class LampHSM {
private:
    std::vector<std::unique_ptr<LampHSMCompartment>> _state_stack;
    std::unique_ptr<LampHSMCompartment> __compartment;
    std::unique_ptr<LampHSMCompartment> __next_compartment;
    std::vector<LampHSMFrameContext> _context_stack;

    void __kernel(LampHSMFrameEvent& __e) {
        __router(__e);
        while (__next_compartment) {
            auto next_compartment = std::move(__next_compartment);
            LampHSMFrameEvent exit_event("<$");
            __router(exit_event);
            __compartment = std::move(next_compartment);
            if (!__compartment->forward_event) {
                LampHSMFrameEvent enter_event("$>");
                __router(enter_event);
            } else {
                auto forward_event = std::move(__compartment->forward_event);
                if (forward_event->_message == "$>") {
                    __router(*forward_event);
                } else {
                    LampHSMFrameEvent enter_event("$>");
                    __router(enter_event);
                    __router(*forward_event);
                }
            }
        }
    }

    void __router(LampHSMFrameEvent& __e) {
        const auto& state_name = __compartment->state;
        if (state_name == "Off") {
            _state_Off(__e);
        } else if (state_name == "On") {
            _state_On(__e);
        } else if (state_name == "ColorBehavior") {
            _state_ColorBehavior(__e);
        }
    }

    void __transition(std::unique_ptr<LampHSMCompartment> next) {
        __next_compartment = std::move(next);
    }

    void _state_ColorBehavior(LampHSMFrameEvent& __e) {
        if (__e._message == "getColor") {
            _context_stack.back()._return = std::any(color);
            return;;
        } else if (__e._message == "setColor") {
            auto color = std::any_cast<std::string>(__e._parameters.at("color"));
            this->color = color;
        }
    }

    void _state_Off(LampHSMFrameEvent& __e) {
        if (__e._message == "isLampOn") {
            _context_stack.back()._return = std::any(lamp_on);
            return;;
        } else if (__e._message == "turnOn") {
            auto __new_compartment = std::make_unique<LampHSMCompartment>("On");
            __new_compartment->parent_compartment = __compartment->clone();
            __transition(std::move(__new_compartment));
            return;
        } else {
            _state_ColorBehavior(__e);
        }
    }

    void _state_On(LampHSMFrameEvent& __e) {
        if (__e._message == "<$") {
            this->turnOffLamp();
        } else if (__e._message == "$>") {
            this->turnOnLamp();
        } else if (__e._message == "isLampOn") {
            _context_stack.back()._return = std::any(lamp_on);
            return;;
        } else if (__e._message == "turnOff") {
            auto __new_compartment = std::make_unique<LampHSMCompartment>("Off");
            __new_compartment->parent_compartment = __compartment->clone();
            __transition(std::move(__new_compartment));
            return;
        } else {
            _state_ColorBehavior(__e);
        }
    }

    void turnOnLamp() {
                    lamp_on = true;
    }

    void turnOffLamp() {
                    lamp_on = false;
    }

public:
    std::string color = "white";
    bool lamp_on = false;

    LampHSM() {
        // HSM: Create parent compartment chain
        auto __parent_comp_0 = std::make_unique<LampHSMCompartment>("ColorBehavior");
        __compartment = std::make_unique<LampHSMCompartment>("Off");
        __compartment->parent_compartment = std::move(__parent_comp_0);
        LampHSMFrameEvent __frame_event("$>");
        LampHSMFrameContext __ctx(std::move(__frame_event));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    void turnOn() {
        LampHSMFrameEvent __e("turnOn");
        LampHSMFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    void turnOff() {
        LampHSMFrameEvent __e("turnOff");
        LampHSMFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    std::string getColor() {
        LampHSMFrameEvent __e("getColor");
        LampHSMFrameContext __ctx(std::move(__e), std::any(std::string()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<std::string>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }

    void setColor(std::string color) {
        std::unordered_map<std::string, std::any> __params;
        __params["color"] = color;
        LampHSMFrameEvent __e("setColor", std::move(__params));
        LampHSMFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    bool isLampOn() {
        LampHSMFrameEvent __e("isLampOn");
        LampHSMFrameContext __ctx(std::move(__e), std::any(bool()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<bool>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }
};

int main() {
    printf("=== Test 32: Doc Lamp HSM ===\n");
    LampHSM lamp;

    if (lamp.getColor() != "white") {
        printf("FAIL: Expected 'white', got '%s'\n", lamp.getColor().c_str());
        assert(false);
    }
    lamp.setColor("red");
    if (lamp.getColor() != "red") {
        printf("FAIL: Expected 'red', got '%s'\n", lamp.getColor().c_str());
        assert(false);
    }

    lamp.turnOn();
    if (lamp.isLampOn() != true) {
        printf("FAIL: Lamp should be on\n");
        assert(false);
    }

    lamp.setColor("green");
    if (lamp.getColor() != "green") {
        printf("FAIL: Expected 'green', got '%s'\n", lamp.getColor().c_str());
        assert(false);
    }

    lamp.turnOff();
    if (lamp.isLampOn() != false) {
        printf("FAIL: Lamp should be off\n");
        assert(false);
    }

    if (lamp.getColor() != "green") {
        printf("FAIL: Expected 'green', got '%s'\n", lamp.getColor().c_str());
        assert(false);
    }

    printf("PASS: HSM lamp works correctly\n");
    return 0;
}
